dissect/target/plugins/os/unix/log/utmp.py

```diff
@@ -68,6 +86,30 @@ def __iter__(self):
 
         while True:
             try:
-                yield utmp.entry(byte_stream)
+                entry = utmp.entry(byte_stream)
+
+                r_type = ""
+                if entry.ut_type in utmp.Type.reverse:
+                    r_type = utmp.Type.reverse[entry.ut_type]
+
+                # UTMP misuses the field ut_addr_v6 for IPv4 and IPv6 addresses, because of this
```

Comment:
Can you write a little test maybe for this fix? So we can easily spot a regression?

Reply:
The fix introduces the same bug in case the IPv6 address is 4 bytes long, but works for IPv6 address that are longer than 4 bytes.

The IPv6 address is `1337:1::'` and in the WTP log is b'\x137\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', the last 12 bytes are zeroes. Because of this IPv6 addresses of which the binary representation is 4 bytes long is parsed as IPv4 address which leads to the address `19.55.0.1`:

```
>>> import ipaddress
>>> ipv6 = "1337:1::"
>>> ipaddress.ip_address(ipv6)
IPv6Address('1337:1::')
>>> ipaddress.ip_address(ipv6).packed
b'\x137\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
>>> ipaddress.ip_address(ipv6).packed[:4]
b'\x137\x00\x01'
>>> ipaddress.ip_address(b"\x137\x00\x01")
IPv4Address('19.55.0.1')
```

I would like to make the suggestion to make the UTMP parser a bit smarter to fix this issue.

* `ut_addr_v6` field uses more then 4 bytes -> IPv6 address
* `ut_addr_v6` field uses 4 bytes
	* If `ut_host is a valid IPv6 address the full 16 bytes of `ut_addr_v6` are parsed as IPv6 address
	* Otherwise the first 4 bytes are parsed as IPv4 address

Reply:
Ah my mistake thanks for seeing that.

I have created a new test file that includes the field `ut->ut_addr_v6`.

As shown in the screenshot utmpdump contains the same issue.

I have used SSH using the IPv6 address `1337::1` which uses 16 bytes, utmpdump correctly shows that the hostname is `1337::1` and the IP-address is `1337::1`.

I have used SSH using the IPv6 address `1337:1::` which uses 4 bytes lead by 12 bytes of trailing zeroes, utmpdump correctly shows that the hostname is `1337:1::` which is correct and the IP-address is `19.55.0.1` which is incorrect. 

![image](https://github.com/fox-it/dissect.target/assets/21525407/39091255-1ad4-4ea3-bcc2-cfb059db4bf9)

The last 12 bytes of leading zeroes causes utmpdump to read the buffer as `AF_INET` instead of `AF_INET6`.

```python
>>> import ipaddress
>>> ipv6 = ipaddress.ip_address("1337:1::")
>>> ipv6
IPv6Address('1337:1::')
>>> ipv6.packed
b'\x137\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
>>> ipv6.packed[:4]
b'\x137\x00\x01'
>>> ipaddress.ip_address(ipv6.packed[:4])
IPv4Address('19.55.0.1')
```

To prevent this issue as shown in the test the `ut_host` field is used to determine the type of the IP-address

Reply:
I don't known how the last case can be reproduced on a Linux machine and correctly parsed by the parser.

| `ut_host` | `ut_addr_v6` | result |
|---|---|---|
| IPv4 address  | IPv4 address |  `ut_addr_v6[0]` is parsed as IPv4 address |
| Random string  | IPv4 address |  `ut_addr_v6[0]` is parsed as IPv4 address |
| Pv6 address  | IPv6 address > 4 bytes)|  `ut_addr_v6` is parsed as IPv6 address |
| Random string  | IPv6 address > 4 bytes)|  `ut_addr_v6` is parsed as IPv6 address |
|  IPv6 address   | IPv6 address <= 4 bytes |  `ut_addr_v6` is parsed as IPv6 address |
|  Random string   | IPv6 address <= 4 bytes |  `ut_addr_v6[0]` is parsed as IPv6 address which is incorrect https://github.com/fox-it/dissect.target/pull/292/files#diff-4676dfeb3cf44a047e82d239cd74ead57f206d968ce3a428c24e4c7b92869272R152-R154 |

Reply:
Ah I see, the utmpdump tool itself also puts out the wrong information. So basically the way the utmp structure is set filled is fundamentally broken unfortunately.

I don't think the final case in your table is easily fixable as there is just too little information. On the plus side I guess that encountering a IP6 address with only the first 32 bits set will probably be rare.